npm/pkg/controlplane/controllers/v2/namespacecontroller.go

```diff
@@ -0,0 +1,459 @@
+			err = nsc.cleanDeletedNamespace(cachedNsKey)
+			if err != nil {
+				// need to retry this cleaning-up process
+				metrics.SendErrorLogAndMetric(util.NSID, "Error: %v when namespace is not found", err)
+				return fmt.Errorf("Error: %v when namespace is not found", err)
+			}
+		}
+		return err
+	}
+
+	if nsObj.DeletionTimestamp != nil || nsObj.DeletionGracePeriodSeconds != nil {
+		return nsc.cleanDeletedNamespace(cachedNsKey)
+	}
+
+	cachedNsObj, nsExists := nsc.npmNamespaceCache.NsMap[cachedNsKey]
+	if nsExists {
+		if reflect.DeepEqual(cachedNsObj.LabelsMap, nsObj.ObjectMeta.Labels) {
+			klog.Infof("[NAMESPACE UPDATE EVENT] Namespace [%s] labels did not change", key)
+			return nil
+		}
+	}
+
+	err = nsc.syncUpdateNamespace(nsObj)
+	if err != nil {
+		metrics.SendErrorLogAndMetric(util.NSID, "[syncNamespace] failed to sync namespace due to  %s", err.Error())
+		return err
+	}
+
+	return nil
+}
+
+// syncAddNamespace handles adding namespace to ipset.
+func (nsc *NamespaceController) syncAddNamespace(nsObj *corev1.Namespace) error {
+	var err error
+	corev1NsName, corev1NsLabels := util.GetNSNameWithPrefix(nsObj.ObjectMeta.Name), nsObj.ObjectMeta.Labels
+	klog.Infof("NAMESPACE CREATING: [%s/%v]", corev1NsName, corev1NsLabels)
+	// Create ipset for the namespace.
+	if err = nsc.dp.CreateIPSet(corev1NsName, ipsets.Namespace); err != nil {
+		metrics.SendErrorLogAndMetric(util.NSID, "[AddNamespace] Error: failed to create ipset for namespace %s with err: %v", corev1NsName, err)
+		return err
+	}
+
+	if err = nsc.dp.AddToList(util.KubeAllNamespacesFlag, []string{corev1NsName}); err != nil {
+		metrics.SendErrorLogAndMetric(util.NSID, "[AddNamespace] Error: failed to add %s to all-namespace ipset list with err: %v", corev1NsName, err)
+		return err
+	}
+
+	npmNs := newNs(corev1NsName)
+	nsc.npmNamespaceCache.NsMap[corev1NsName] = npmNs
+
+	// Add the namespace to its label's ipset list.
+	for nsLabelKey, nsLabelVal := range corev1NsLabels {
+		labelIpsetName := util.GetNSNameWithPrefix(nsLabelKey)
+		klog.Infof("Adding namespace %s to ipset list %s", corev1NsName, labelIpsetName)
+		if err = nsc.dp.AddToList(labelIpsetName, []string{corev1NsName}); err != nil {
+			metrics.SendErrorLogAndMetric(util.NSID, "[AddNamespace] Error: failed to add namespace %s to ipset list %s with err: %v", corev1NsName, labelIpsetName, err)
+			return err
+		}
+
+		labelIpsetName = util.GetNSNameWithPrefix(util.GetIpSetFromLabelKV(nsLabelKey, nsLabelVal))
+		klog.Infof("Adding namespace %s to ipset list %s", corev1NsName, labelIpsetName)
+		if err = nsc.dp.AddToList(labelIpsetName, []string{corev1NsName}); err != nil {
+			metrics.SendErrorLogAndMetric(util.NSID, "[AddNamespace] Error: failed to add namespace %s to ipset list %s with err: %v", corev1NsName, labelIpsetName, err)
+			return err
+		}
+
+		// Append succeeded labels to the cache NS obj
+		npmNs.appendLabels(map[string]string{nsLabelKey: nsLabelVal}, AppendToExistingLabels)
+	}
+
```

Comment:
At the end of all these updates to IpSets, in NS controller we will need to call "dp.ApplyDataPlane()" this calculates needed updates and applies them, until then all the changes will be in only staged in cache.

Reply:
This ApplyDataplane call should be done at the end of all 3 events for both controllers